apps/builder/app/builder/features/style-panel/sections/backgrounds/background-layers.ts

```diff
@@ -0,0 +1,254 @@
+};
+
+const getPropertyLayerCount = (style: StyleValueInfo) => {
+  if (style.value.type === "layers") {
+    return style.value.value.length;
+  }
+  return 0;
+};
+
+export const getLayerCount = (style: StyleInfo) => {
+  return Math.max(
+    ...layeredBackgroundProps.map((prop) => {
+      const styleValue = style[prop];
+
+      if (styleValue !== undefined) {
+        return getPropertyLayerCount(styleValue);
+      }
+
+      return 0;
+    })
+  );
+};
+
+export const getLayerBackgroundStyleInfo = (
+  layerNum: number,
+  style: StyleInfo
+): StyleInfo => {
+  const layerCount = getLayerCount(style);
+  if (layerNum >= layerCount) {
+    throw new Error(`${layerNum} is out of bounds`);
+  }
+
+  const result: StyleInfo = {};
+
+  for (const [property, value] of Object.entries(
+    layeredBackgroundPropsDefaults
+  )) {
+    result[property as LayeredBackgroundProperty] = { value };
+  }
+
+  for (const property of layeredBackgroundProps) {
+    const resultProperty = result[property];
+    if (resultProperty === undefined) {
+      throw new Error(`Property ${property} is not defined`);
+    }
+
+    const styleValue = style[property];
+
+    const valueStyle = styleValue?.value;
+    const localStyle = styleValue?.local;
+    const cascadedStyle = styleValue?.cascaded;
+
+    if (valueStyle?.type === "layers") {
+      const styleValue = valueStyle.value[layerNum];
+      resultProperty["value"] = styleValue;
+    }
+    if (localStyle?.type === "layers") {
+      const styleValue = localStyle.value[layerNum];
+      resultProperty["local"] = styleValue;
+    }
+    if (cascadedStyle?.value.type === "layers") {
+      const styleValue = cascadedStyle.value.value[layerNum];
+      resultProperty["cascaded"] = {
+        breakpointId: cascadedStyle.breakpointId,
+        value: styleValue,
+      };
+    }
+  }
+
+  return result;
+};
+
+export type DeleteBackgroundProperty = ReturnType<typeof deleteLayerProperty>;
+
+export const deleteLayerProperty =
+  (
+    layerNum: number,
+    style: StyleInfo,
+    deleteProperty: DeleteProperty,
+    createBatchUpdate: CreateBatchUpdate
+  ) =>
+  (propertyName: LayeredBackgroundProperty, options?: StyleUpdateOptions) => {
+    if (options?.isEphemeral) {
+      deleteProperty(propertyName, options);
+      return;
+    }
+
+    setLayerProperty(layerNum, style, createBatchUpdate)(propertyName)(
+      layeredBackgroundPropsDefaults[propertyName]
+    );
+  };
+
+export type SetBackgroundProperty = ReturnType<typeof setLayerProperty>;
+
+export const setLayerProperty =
+  (layerNum: number, style: StyleInfo, createBatchUpdate: CreateBatchUpdate) =>
+  (propertyName: LayeredBackgroundProperty) =>
+  (newValue: BackgroundStyleValue, options?: StyleUpdateOptions) => {
+    const batch = createBatchUpdate();
+
+    const layerCount = Math.max(getLayerCount(style), layerNum + 1);
+
+    for (const property of layeredBackgroundProps) {
+      let propertyChanged = false;
```

Comment:
this style with a flag that changes depending on conditions in a long block is generally problematic: hard to read, easy to break,
much better would be an early return for each case